[libc++] Implement part of P2562R1: constexpr `std::inplace_merge`

frederick-vs-ja opened the pull request:
Fixes #119398.

Drive-by:
- Adds `constexpr_random.h` for pseudo-randomizing or shuffling in tests for constant evaluation.

frederick-vs-ja committed:
[libc++] Implement constexpr `std::inplace_merge`

frederick-vs-ja committed:
Clang-format some test files

mordante commented on libcxx/include/__algorithm/inplace_merge.h:
```diff
@@ -223,14 +224,14 @@ _LIBCPP_HIDE_FROM_ABI void __inplace_merge(
 
 template <class _BidirectionalIterator, class _Compare>
-inline _LIBCPP_HIDE_FROM_ABI void inplace_merge(
+_LIBCPP_HIDE_FROM_ABI inline _LIBCPP_CONSTEXPR_SINCE_CXX26 void inplace_merge(
```
`inline` first seems to be the most common in the code base, so let's keep it like that.
```suggestion
inline _LIBCPP_HIDE_FROM_ABI  _LIBCPP_CONSTEXPR_SINCE_CXX26 void inplace_merge(
```

mordante commented on libcxx/include/algorithm:
```diff
@@ -1606,11 +1606,11 @@ template <class InputIterator1, class InputIterator2, class OutputIterator, clas
 
 template <class BidirectionalIterator>
-    void
+    constexpr void                                    // constexpr in C++26
```
```suggestion
    constexpr void                                    // constexpr since C++26
```

mordante commented on libcxx/test/std/algorithms/alg.sorting/alg.merge/inplace_merge.pass.cpp:
```diff
@@ -25,99 +23,146 @@
+}
+
+template <class Iter>
+TEST_CONSTEXPR_CXX26 void test_one_non_randomized(unsigned N, unsigned M) {
```
The name of this function looks odd; it looks like you're using a simple random engine.

Is there a reason why we can't remove the old function and use this one instead?
If we can't we should document why not.

frederick-vs-ja replied:
Actually, I think it would make more sense to only keep the old function while making it constant-evaluation-compatible. Perhaps we can create `constexpr` versions for `std::mt19937` and `std::shuffle` and use them in constant evaluation.

mordante replied:
I think it makes sense to have a `constexpr` random generator for tests. I would look at a less complex engine than `std::mt19937`. I don't think we need a very strong generator and it's good to keep the number of operations lower during constant evaluation. WDYT?

frederick-vs-ja replied:
Agreed. I'll use constexpr-ized version of `minstd_rand`.

mordante commented on libcxx/test/std/algorithms/alg.sorting/alg.merge/inplace_merge.pass.cpp:
```diff
@@ -25,99 +23,146 @@
+  std::sort(ia + M, ia + N);
+  std::inplace_merge(Iter(ia), Iter(ia + M), Iter(ia + N));
+  if (N > 0) {
+    assert(std::is_sorted(ia, ia + N));
```
This looks odd; when N == 0 we execute code, but nothing is tested. Do you know why this is useful? 
(The original code does the same.)

frederick-vs-ja replied:
> Do you know why this is useful?

No, I just copied `if (N > 0) {` from the old code. Perhaps this condition should be removed.

mordante commented on libcxx/test/support/counting_predicates.h:
```diff
@@ -41,14 +41,17 @@ struct binary_counting_predicate {
-    binary_counting_predicate ( Predicate p ) : p_(p), count_(0) {}
-    ~binary_counting_predicate() {}
+    TEST_CONSTEXPR_CXX20 binary_counting_predicate(Predicate p) : p_(p), count_(0) {}
+    TEST_CONSTEXPR_CXX20 ~binary_counting_predicate() {}
```
Doesn't this work?
```suggestion
    ~binary_counting_predicate() = delete;
```

frederick-vs-ja replied:
This doesn't work because we need to construct `binary_counting_predicate` objects.

IIUC implicitly declared (defaulted) destructor should works. Although I'm not sure whether it was intended not to use defaulted destructors.

mordante replied:
Sorry I meant `= default;`, not `= delete;`. The reason why it was not done in the past since `= default` is not part of C++03. Nowadays Clang does support that syntax in C++03.

frederick-vs-ja replied:
I'm switching to use implicit declared destructor, which is also defaulted and possibly trivial in C++03.

mordante reviewed (commented):
Thanks for working on this! Some comments regarding the tests.

frederick-vs-ja committed:
Merge branch 'main' into constexpr-std-inplace_merge

frederick-vs-ja committed:
Change `constexpr in` to `constexpr since` in comments

frederick-vs-ja committed:
Add `constexpr_random.h` for test suite

frederick-vs-ja committed:
Use constexpr-ized `minstd_rand` as random source in test files

frederick-vs-ja committed:
Implicitly declare destructor for `binary_counting_predicate`

frederick-vs-ja committed:
Reorder `inline` in production code

frederick-vs-ja committed:
Clang-format `constexpr_random.h`

frederick-vs-ja committed:
Merge branch 'main' into constexpr-std-inplace_merge

frederick-vs-ja committed:
Include `<limits>` in `constexpr_random.h`

frederick-vs-ja committed:
Merge branch 'main' into constexpr-std-inplace_merge

frederick-vs-ja committed:
Fix mis-versioned `constexpr` in `constexpr_random.h`

mordante commented on libcxx/test/std/algorithms/alg.sorting/alg.merge/inplace_merge.pass.cpp:
```diff
@@ -8,116 +8,135 @@
+  test<Iter>(4, randomness);
+  test<Iter>(50, randomness);
+#if defined(_LIBCPP_HARDENING_MODE)
+  if (!TEST_IS_CONSTANT_EVALUATED) // avoid blowing past constant evaluation limit
```
I would like to use a bit "nicer" language.
```suggestion
  if (!TEST_IS_CONSTANT_EVALUATED) // avoid exceeding the constant evaluation step limit
```

mordante commented on libcxx/test/support/constexpr_random.h:
```diff
@@ -0,0 +1,705 @@
+//===----------------------------------------------------------------------===//
+
+#ifndef TEST_SUPPORT_CONSTEXPR_RANDOM_H
+#define TEST_SUPPORT_CONSTEXPR_RANDOM_H
```
I feel this is quite a lot of code. I know we've had bugs in libc++ with Schrage's algorithm. So I'm a bit reluctant to add this code without any tests; probably in a separate commit. I still wonder whether we shouldn't add a simplified random algorithm. Not only to keep the code simple, but also to keep the number of steps during constant evaluation lower. WDYT?

frederick-vs-ja replied:
Agreed. I'll look for a simpler RNG.

mordante replied:
Thanks. It's still a bit more code than I hoped, but I don't think that can be avoided.

frederick-vs-ja committed:
Merge branch 'main' into constexpr-std-inplace_merge

frederick-vs-ja committed:
Address @mordante's review comments

- Uses a simpler RNG.
- Improves comments.

frederick-vs-ja committed:
Fix errors in C++03

frederick-vs-ja committed:
Merge branch 'main' into constexpr-std-inplace_merge

frederick-vs-ja committed:
Restore skipping

mordante commented on libcxx/test/support/constexpr_random.h:
```diff
@@ -0,0 +1,438 @@
+#endif
+  TEST_CONSTEXPR explicit simple_random_generator(std::uint16_t s) TEST_NOEXCEPT : status_(s) {}
+
+  TEST_CONSTEXPR_CXX14 result_type operator()() TEST_NOEXCEPT {
```
Can you add a comment which engine this is and which values values?
For example
`A linear congruential generator, using the constants used by MSVC.`

frederick-vs-ja replied:
Comment added - `// A linear congruential generator, using the constants used by MS UCRT.`

mordante commented on libcxx/test/std/algorithms/alg.sorting/alg.merge/inplace_merge.pass.cpp:
```diff
@@ -8,116 +8,135 @@
+  test_one<Iter>(3, 3, randomness);
+  test<Iter>(4, randomness);
+  test<Iter>(50, randomness);
+#if defined(_LIBCPP_HARDENING_MODE)
+  if (!TEST_IS_CONSTANT_EVALUATED) // avoid exceeding the constant evaluation step limit
+#endif
+  {
+    test<Iter>(100, randomness);
+  }
+  if (!TEST_IS_CONSTANT_EVALUATED) { // avoid exceeding the constant evaluation step limit
+    test<Iter>(1000, randomness);
+  }
```
I wonder how useful 100 and 1000. Let's at least disable them both unconditionally during constant evaluation.
That simplifies the code by removing the `_LIBCPP_HARDENING_MODE` ifdef.

We try to avoid using `_LIBCPP_FOO` macros in our tests.

mordante commented on libcxx/test/std/algorithms/alg.sorting/alg.merge/inplace_merge_comp.pass.cpp:
```diff
@@ -8,168 +8,183 @@
+  test<Iter>(4, randomness);
+  test<Iter>(20, randomness);
+  test<Iter>(50, randomness);
+#if defined(_LIBCPP_HARDENING_MODE)
+  if (!TEST_IS_CONSTANT_EVALUATED) // avoid exceeding the constant evaluation step limit
+#endif
+  {
+    test<Iter>(100, randomness);
+  }
+  if (!TEST_IS_CONSTANT_EVALUATED) { // avoid exceeding the constant evaluation step limit
+    test<Iter>(1000, randomness);
+  }
```
Same here.

frederick-vs-ja replied:
Done.

mordante reviewed (approved):
Thanks a few minor issues, otherwise LGTM!

frederick-vs-ja committed:
Merge branch 'main' into constexpr-std-inplace_merge

frederick-vs-ja committed:
Address @mordante's review comments

- Skip 100 cases for constant evaluation.
- Cite the source of MS UCRT's constants.

mordante reviewed (approved):
LGTM!

ldionne commented on libcxx/test/std/algorithms/alg.sorting/alg.merge/inplace_merge.pass.cpp:
```diff
@@ -8,116 +8,130 @@
-    test<random_access_iterator<int*> >();
-    test<int*>();
+TEST_CONSTEXPR_CXX26 bool test() {
+  support::simple_random_generator randomness;
```
FWIW, I think I would have favoured an approach where we get rid of the randomness in these tests instead, but I think this is acceptable and certainly has benefits. No action needed.
